src/Http/Routing/src/DefaultLinkGenerator.cs

```diff
@@ -336,137 +336,92 @@ public void Dispose()
-                }
-            }
+            [LoggerMessage(EventId = 100, EventName = "EndpointsFound", Level = LogLevel.Debug, Message = "Found the endpoints {Endpoints} for address {Address}")]
+            public static partial void EndpointsFound(ILogger logger, object address, IEnumerable<Endpoint> endpoints);
 
-            public static void EndpointsNotFound(ILogger logger, object address)
-            {
-                _endpointsNotFound(logger, address, null);
-            }
+            [LoggerMessage(EventId = 101, EventName = "EndpointsNotFound", Level = LogLevel.Debug, Message = "No endpoints found for address {Address}")]
+            public static partial void EndpointsNotFound(ILogger logger, object address);
 
             public static void TemplateSucceeded(ILogger logger, RouteEndpoint endpoint, PathString path, QueryString query)
-            {
-                _templateSucceeded(logger, endpoint.RoutePattern.RawText, endpoint.DisplayName, path.Value, query.Value, null);
-            }
+                => TemplateSucceeded(logger, endpoint.RoutePattern.RawText, endpoint.DisplayName, path.Value, query.Value);
+
+            [LoggerMessage(EventId = 102, EventName = "TemplateSucceeded", Level = LogLevel.Debug,
+                Message = "Successfully processed template {Template} for {Endpoint} resulting in {Path} and {Query}")]
+            public static partial void TemplateSucceeded(ILogger logger, string template, string endpoint, string path, string query);
 
             public static void TemplateFailedRequiredValues(ILogger logger, RouteEndpoint endpoint, RouteValueDictionary ambientValues, RouteValueDictionary values)
             {
                 // Checking level again to avoid allocation on the common path
                 if (logger.IsEnabled(LogLevel.Debug))
                 {
-                    _templateFailedRequiredValues(logger, endpoint.RoutePattern.RawText, endpoint.DisplayName, FormatRouteValues(ambientValues), FormatRouteValues(values), FormatRouteValues(endpoint.RoutePattern.Defaults), null);
+                    TemplateFailedRequiredValues(logger, endpoint.RoutePattern.RawText, endpoint.DisplayName, FormatRouteValues(ambientValues), FormatRouteValues(values), FormatRouteValues(endpoint.RoutePattern.Defaults));
                 }
             }
 
+            [LoggerMessage(EventId = 103, EventName = "TemplateFailedRequiredValues", Level = LogLevel.Debug,
+                Message = "Failed to process the template {Template} for {Endpoint}. " +
+                "A required route value is missing, or has a different value from the required default values. " +
+                "Supplied ambient values {AmbientValues} and {Values} with default values {Defaults}")]
+            public static partial void TemplateFailedRequiredValues(ILogger logger, string template, string endpoint, string ambientValues, string values, string defaults);
```

Comment:
Do partial methods on log generators always have to be public? Looks like this would only be called from the overload so it could be private.